File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseMatch.java
package com.yahyeet.boardbook.model.firebase.repository;

import com.google.firebase.firestore.DocumentSnapshot;
import com.yahyeet.boardbook.model.entity.Match;

import java.util.HashMap;
import java.util.Map;

public class FirebaseMatch extends FirebaseEntity<Match> {
	private String id;
	private String gameId;

	public FirebaseMatch() {
	}

	public FirebaseMatch(String id, String gameId) {
		this.id = id;
		this.gameId = gameId;
	}

	@Override
	public Map<String, Object> toMap() {
		Map<String, Object> map = new HashMap<>();

		if (gameId != null) {
			map.put("gameId", gameId);
		}

		return map;
	}

	@Override
	public Match toModelType() {
		return new Match(id);
	}

	public static FirebaseMatch fromModelType(Match match) {
		return new FirebaseMatch(match.getId(), match.getGame().getId());
	}

	public static FirebaseMatch fromDocument(DocumentSnapshot document) {
		FirebaseMatch firebaseMatch = new FirebaseMatch();
		firebaseMatch.setId(document.getId());

		if (document.contains("gameId")) {
			firebaseMatch.setGameId(document.getString("gameId"));
		}

		return firebaseMatch;
	}

	public String getId() {
		return id;
	}

	public void setId(String id) {
		this.id = id;
	}

	public String getGameId() {
		return gameId;
	}

	public void setGameId(String gameId) {
		this.gameId = gameId;
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseMatch.kt
package com.yahyeet.boardbook.model.firebase.repository

import com.google.firebase.firestore.DocumentSnapshot
import com.yahyeet.boardbook.model.entity.Game
import com.yahyeet.boardbook.model.entity.Match

import java.util.HashMap

class FirebaseMatch : AbstractFirebaseEntity<Match> {
    var gameId: String? = null

    constructor() {}

    constructor(id: String, gameId: String) : super(id) {
        this.gameId = gameId
    }

    override fun toMap(): Map<String, Any> {
        val map = HashMap<String, Any>()

        if (gameId != null) {
            map["gameId"] = gameId!!
        }

        return map
    }

    override fun toModelType(): Match {
        val match = Match(id)

        match.game = Game(gameId)

        return match
    }

    companion object {

        fun fromModelType(match: Match): FirebaseMatch {
            return FirebaseMatch(match.id, match.game!!.id)
        }

        fun fromDocument(document: DocumentSnapshot): FirebaseMatch {
            val firebaseMatch = FirebaseMatch()
            firebaseMatch.id = document.id

            if (document.contains("gameId")) {
                firebaseMatch.gameId = document.getString("gameId")
            }

            return firebaseMatch
        }
    }
}
